### PR TITLE
modify xgboost with prediction for other data frame

### PR DESCRIPTION
**Pull request by double-y**
### Description
modify xgboost with prediction for other data frame

https://github.com/exploratory-io/tam/issues/5145 (only for xgboost)

### Checklist

Make sure you have performed following items before submitting this pull request.
If not, please describe the reason.  

- [x] Add test cases for this fix/enhancement
- [x] Pass devtools::check()
- [x] Pass devtools::test()
- [ ] Test installing from github
- [x] Tested with Exploratory

### CODE REVIEW

File: R/broom_wrapper.R [hunk 1]
```diff
@@ -242,136 +246,143 @@ kmeans_info <- function(df){
...
-        safe_slice(df, index, remove = TRUE)
-      })) %>%
-      dplyr::select(-.test_index) %>%
-      dplyr::rowwise() %>%
-      # evaluate the formula of augment and "data" column will have it
-      dplyr::mutate_(.dots = list(source.data = aug_fml))
-
-    if (with_response){
-      augmented <- augmented %>%
+      # augment by trainig data
```
Comment (hideaki) [inline R/broom_wrapper.R]: typo: trainig

File: R/broom_wrapper.R [hunk 2]
```diff
@@ -82,28 +82,41 @@ augment_kmeans <- function(df, model, data){
...
       # .test_index is used because model_df has it and won't be used here
       dplyr::mutate_(.dots = list(.test_index = aug_fml)) %>%
-      dplyr::ungroup() %>%
-      dplyr::mutate(.test_index = purrr::map2(.test_index, model, function(d, m){
-        # add predicted_response to the result data frame
-        add_response(d, m, "predicted_response")
-      })) %>%
+      dplyr::ungroup()
+    ret <- if(with_respose) {
+     ret %>%
```
Comment (hideaki) [inline R/broom_wrapper.R]: indent is off here.

File: R/broom_wrapper.R [hunk 3]
```diff
@@ -242,136 +246,143 @@ kmeans_info <- function(df){
 
 #' augment using source data and test index
 #' @param df Data frame that has model and .test_index.
-#' @param data "training" or "test". Which source data should be used.
+#' @param data "training" or "test" or "df". Which source data should be used.
```
Comment (hideaki) [inline R/broom_wrapper.R]: "df" sounds kind of vague. How about "newdata" like broom?

File: R/build_xgboost.R [hunk 4]
```diff
@@ -35,6 +35,13 @@ fml_xgboost <- function(data, formula, nrounds= 10, weights = NULL, watchlist_ra
     }
     ret$terms <- term
     ret$x_names <- colnames(md_mat)
+    pred_cnames <- all.vars(term)[-1]
+    types <- vapply(pred_cnames, function(cname) {
+      get_data_type(md_frame[[cname]])
+    }, FUN.VALUE = "")
+    names(types) <- pred_cnames
+    ret$types <- types
+    ret$xlevels <- .getXlevels(term, md_frame)
```
Comment (hideaki) [inline R/build_xgboost.R]: explain what xlevels is.

File: R/build_xgboost.R [hunk 5]
```diff
@@ -351,11 +358,41 @@ augment.xgboost_reg <- function(x, data = NULL, newdata = NULL, ...) {
       data
     }
 
-    mat_data <- model.matrix(x$terms, data = ret_data)
+    # validate column types
```
Comment (hideaki) [inline R/build_xgboost.R]: explain this more.

File: R/build_xgboost.R [hunk 6]
```diff
@@ -351,11 +358,41 @@ augment.xgboost_reg <- function(x, data = NULL, newdata = NULL, ...) {
...
+      }, FUN.VALUE = "")
+
+      if(any(!is.na(message))){
+        stop(paste0("Found data type mismatch: ", paste0(message[!is.na(message)], collapse = ", ")))
+      }
+    }
+
+    y_name <- all.vars(x$terms)[[1]]
+    if(is.null(ret_data[[y_name]])){
+      # if there are no column in the formula, model.matrix causes an error
```
Comment (hideaki) [inline R/build_xgboost.R]: explain the background more.

File: R/build_xgboost.R [hunk 7]
```diff
@@ -351,11 +358,41 @@ augment.xgboost_reg <- function(x, data = NULL, newdata = NULL, ...) {
...
+      # so create the column with 0
+      ret_data[[y_name]] <- rep(0, nrow(ret_data))
+    }
+
+    mat_data <- model.matrix(x$terms, data = ret_data, xlev = x$xlevels)
 
     predicted <- stats::predict(x, mat_data)
     predicted_value_col <- avoid_conflict(colnames(ret_data), "predicted_value")
-    ret_data[[predicted_value_col]] <- predicted
+    ret_data[[predicted_value_col]] <- fill_vec_NA(as.integer(rownames(mat_data)), predicted, nrow(ret_data))
```
Comment (hideaki) [inline R/build_xgboost.R]: explain this more.

File: R/broom_wrapper.R (hunk 2 above)
Comment (double-y) [inline R/broom_wrapper.R]: I think this is correct. Will add a line between l102 and l103

File: R/build_xgboost.R (hunk 5 above)
Comment (double-y) [inline R/build_xgboost.R]: added comments

File: R/build_xgboost.R (hunk 7 above)
Comment (double-y) [inline R/build_xgboost.R]: added comments

File: R/build_xgboost.R (hunk 4 above)
Comment (double-y) [inline R/build_xgboost.R]: added a comment

File: R/broom_wrapper.R (hunk 1 above)
Comment (double-y) [inline R/broom_wrapper.R]: fixed it

File: R/build_xgboost.R (hunk 6 above)
Comment (double-y) [inline R/build_xgboost.R]: added comments

File: R/broom_wrapper.R (hunk 2 above)
Comment (hideaki) [inline R/broom_wrapper.R]: The following line has only one space for its indentation. Is this intentional?
ret %>%

Comment (double-y) [inline R/broom_wrapper.R]: Oh, I see. That's right. will fix it

**Pull request merged**